src/agent/rustjail/src/cgroups/systemd/manager.rs

```diff
@@ -0,0 +1,129 @@
+    dbus_client: DBusClient,
+    // fs manager for get properties
+    fs_manager: FsManager,
+    // cgroup version for different dbus properties
+    cg_hierarchy: CgroupHierarchy,
+}
+
+impl CgroupManager for Manager {
+    fn apply(&self, pid: pid_t) -> Result<()> {
+        let unit_name = self.unit_name.as_str();
+        if self.dbus_client.unit_exist(unit_name).unwrap() {
+            self.dbus_client.add_process(pid, self.unit_name.as_str())?;
+        } else {
+            self.dbus_client.start_unit(
+                (pid as u32).try_into().unwrap(),
+                self.cgroups_path.slice.as_str(),
+                self.unit_name.as_str(),
+                &self.cg_hierarchy,
+            )?;
+        }
+
+        Ok(())
+    }
+
+    fn set(&self, r: &LinuxResources, _: bool) -> Result<()> {
+        let mut properties: Properties = vec![];
+
+        let systemd_version = self.dbus_client.get_version()?;
+        let systemd_version_str = systemd_version.as_str();
+
+        Cpu::apply(r, &mut properties, &self.cg_hierarchy, systemd_version_str)?;
+        Memory::apply(r, &mut properties, &self.cg_hierarchy, systemd_version_str)?;
+        Pids::apply(r, &mut properties, &self.cg_hierarchy, systemd_version_str)?;
+        CpuSet::apply(r, &mut properties, &self.cg_hierarchy, systemd_version_str)?;
+
+        self.dbus_client
+            .set_properties(self.unit_name.as_str(), &properties)?;
+
+        Ok(())
+    }
+
+    fn get_stats(&self) -> Result<CgroupStats> {
+        self.fs_manager.get_stats()
+    }
+
+    fn freeze(&self, state: FreezerState) -> Result<()> {
+        self.fs_manager.freeze(state)
+    }
+
+    fn destroy(&mut self) -> Result<()> {
+        self.dbus_client.stop_unit(self.unit_name.as_str())?;
+        self.fs_manager.destroy()
+    }
+
+    fn get_pids(&self) -> Result<Vec<pid_t>> {
+        self.fs_manager.get_pids()
+    }
+
+    fn update_cpuset_path(&self, guest_cpuset: &str, container_cpuset: &str) -> Result<()> {
+        self.fs_manager
+            .update_cpuset_path(guest_cpuset, container_cpuset)
+    }
+
+    fn get_cg_path(&self, cg: &str) -> Result<String> {
+        self.fs_manager.get_cg_path(cg)
+    }
+
+    fn as_any(&self) -> Result<&dyn Any> {
+        Ok(self)
+    }
+}
+
+impl Manager {
+    pub fn new(cgroups_path_str: &str) -> Result<Self> {
+        let cgroups_path = CgroupsPath::new(cgroups_path_str)?;
+        let (parent_slice, unit_name) = cgroups_path.parse()?;
+        let cpath = parent_slice + "/" + &unit_name;
+
+        let fs_manager = FsManager::new(cpath.as_str())?;
+
+        Ok(Manager {
+            paths: fs_manager.paths.clone(),
+            mounts: fs_manager.mounts.clone(),
+            cgroups_path,
+            cpath: cpath.to_owned(),
+            unit_name: unit_name.to_owned(),
+            dbus_client: DBusClient {},
+            fs_manager,
+            cg_hierarchy: if cgroups::hierarchies::is_cgroup2_unified_mode() {
```

Comment:
One property of cgroups v2 is that it can be mixed with v1. It seems that we do not support such a setup. If so, please document it clearly in the design doc.

Reply:
Fixed